internal/jobmanager/jobmanager.go

```diff
@@ -0,0 +1,200 @@
+	if workFunc == nil {
+		return nil, len(workItems), 0, fmt.Errorf("Cannot run job with a nil work function.")
+	}
+
+	err := options.Validate()
+	if err != nil {
+		return nil, len(workItems), 0, fmt.Errorf("Failed to validate job options: '%v'.", err)
+	}
+
+	completeItems := make([]OutputType, 0)
+	remainingItems := workItems
+
+	// If we are not overwriting the cache, query for cached results.
+	if !options.OverwriteCache && cacheFunc != nil {
+		completeItems, remainingItems, err = cacheFunc(workItems)
+		if err != nil {
+			return nil, len(workItems), 0, err
+		}
+	}
+
+	runTime := int64(0)
+	results := make([]OutputType, 0)
+
+	if options.WaitForCompletion {
+		results, runTime, err = runJob(options, remainingItems, cacheFunc, removeCacheFunc, workFunc)
+		if err != nil {
+			return nil, len(workItems), 0, err
+		}
+
+		// If the context was canceled during execution, return immediately.
+		if options.Context.Err() != nil {
+			return nil, 0, 0, nil
+		}
+
+		completeItems = append(completeItems, results...)
+		remainingItems = nil
+	} else {
+		go func() {
+			_, _, err := runJob(options, remainingItems, cacheFunc, removeCacheFunc, workFunc)
+			if err != nil {
+				log.Error("Failure while running asynchronous job.")
+			}
+		}()
+	}
+
+	return completeItems, len(remainingItems), runTime, nil
+}
+
+func runJob[InputType any, OutputType any](options *JobOptions, workItems []InputType, cacheFunc func([]InputType) ([]OutputType, []InputType, error), removeCacheFunc func([]InputType) error, workFunc func(InputType) (OutputType, int64, error)) ([]OutputType, int64, error) {
+	if len(workItems) == 0 {
+		return nil, 0, nil
+	}
+
+	err := options.Validate()
+	if err != nil {
+		return []OutputType{}, 0, fmt.Errorf("Failed to validate job options: '%v'.", err)
+	}
+
+	noLockWait := lockmanager.Lock(options.LockKey)
+	defer lockmanager.Unlock(options.LockKey)
+
+	// The context has been canceled while waiting for a lock, abandon this job.
+	if options.Context.Err() != nil {
+		return nil, 0, nil
+	}
+
+	results := make([]OutputType, 0, len(workItems))
+
+	// If we had to wait for the lock, then check again for cached results.
+	// If there are multiple requests queued up,
+	// it will be faster to do a bulk check for cached results instead of checking each one individually.
+	if !noLockWait && cacheFunc != nil {
+		var partialResults []OutputType = nil
+		partialResults, workItems, err = cacheFunc(workItems)
+		if err != nil {
+			return nil, 0, fmt.Errorf("Failed to re-check result cache before run: '%w'.", err)
+		}
+
+		// Collect the partial results from the cache.
+		results = append(results, partialResults...)
+	}
+
+	if len(workItems) == 0 {
+		return results, 0, nil
+	}
+
+	// If we are overwriting the cache, then remove all the old entries.
+	if options.OverwriteCache && removeCacheFunc != nil && !options.DryRun {
```

Comment:
I think the semantics here may be a bit off (and they are questionable in analysis).
We just checked the cache (and we checked even before the lock),
but now we are removing `workItems`, which is only the non-cached results.
So I think we may have not overwritten the cache.

Note that analysis may still be wrong, but it is a little different since `options.OverwriteCache` is checked in their caching function:
https://github.com/edulinq/autograder-server/blob/main/internal/analysis/individual.go#L62

A warning here:
Do not worry about multiple calls (e.g. to analysis) overwriting each other in the cache.
Because of the lock they will run serially, and they should always give the same result with the same input.